asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py

```diff
@@ -0,0 +1,367 @@
+    min_topic_size = model_configs.get("min_topic_size", 10)
+    vectorizer_model = model_configs.get("vectorizer_model", None)
+    umap_model = model_configs.get("umap_model", None)
+    hdbscan_model = model_configs.get("hdbscan_model", None)
+    embedding_model = model_configs.get("embedding_model", None)
+
+    for i in range(n_runs):
+        topic_model = BERTopic(
+            nr_topics=nr_topics,
+            representation_model=representation_model,
+            min_topic_size=min_topic_size,
+            vectorizer_model=vectorizer_model,
+            umap_model=umap_model,
+            hdbscan_model=hdbscan_model,
+            embedding_model=embedding_model,
+        )
+
+        try:
+            topics, probs = topic_model.fit_transform(docs)
+            topics = topic_model.get_topics()
+
+            topics_info = topic_model.get_topic_info()
+
+            doc_info = topic_model.get_document_info(docs)
+
+            # Append results from each run
+            runs_number_of_topics.append(len(topics))
+            runs_number_of_outliers.append(
+                topics_info[topics_info["Topic"] == -1]["Count"].iloc[0]
+            )
+            # Appending average probability of the documents (not in the outliers' cluster) belonging to topics
+            runs_avg_prob.append(doc_info[doc_info["Topic"] > -1]["Probability"].mean())
+
+            # Updating number of times a doc is an outlier
+            docs_outlier_count = docs_outlier_count.merge(
+                right=doc_info[["Document", "Topic"]], on="Document"
+            )
+            docs_outlier_count["outlier_count"] = docs_outlier_count.apply(
+                lambda x: (
+                    x["outlier_count"] + 1 if x["Topic"] == -1 else x["outlier_count"]
+                ),
+                axis=1,
+            )
+            docs_outlier_count.drop(columns="Topic", inplace=True)
+        except:
+            print(f"Run {i} failed due to no topics being found. Skipping...")
+
+        outliers_vs_docs = (
+            docs_outlier_count.groupby("outlier_count")
+            .size()
+            .reset_index(name="num_docs")
+        )
+        outliers_vs_docs = outliers_vs_docs[outliers_vs_docs["outlier_count"] > 0]
+
+    return (
+        runs_number_of_topics,
+        runs_number_of_outliers,
+        runs_avg_prob,
+        outliers_vs_docs,
+    )
+
+
+def read_and_filter_data(
+    source_name: str, category: str, path_to_data_file: str, keywords: list
+) -> pd.DataFrame:
+    """
+    Loads and filters data before applying topic analysis.
+
+    Args:
+        source_name (str): source name e.g. mse or bh
+        category (str): category e.g. "all"
+        path_to_data_file (str): path to data file if not the standard forum data
+        keywords (list): list of keywords to filter the data
+
+    Returns:
+        pd.DataFrame: filtered data
+    """
+    if path_to_data_file is not None:
```

Comment:
Is the "path_to_data_file" for some other form of data (that's not "mse" or "bh")?

Reply:
yes, as written in the script docstring at the top of the file, it is used "if not reading standard forum data (e.g. if reading questions data)".
So if we want to run this on questions data, we need to pass the path to the questions data file. Let me know if you want me to add more context to this.

Also, we can change it to make it easier to read in questions' data. This was just a simple implementation I thought of, so that you could try and apply this to the questions data if you wanted to.

Reply:
Very well, sounds like a good idea.